Custom error documentation (#10937)
ethereum/solidity: docs/control-structures.rst

    if the account called is non-existent, as part of the design
    of the EVM. Account existence must be checked prior to calling if needed.

Exceptions can contain error data that is passed back to the caller
in the form of :ref:`error instances <errors>`.
The built-in errors ``Error(string)`` and ``Panic(uint256)`` are

> used by special functions, as explained below. The first ("error") is used for "regular" error conditions

Reviewer: Shouldn't it be `The first ("Error(string)") is used..` here? Also the next line should be `while the second "Panic(uint256)" is used..` I think it makes it more clear.

Reviewer 2: I'd say just
```markdown
`Error` is used for "regular" error conditions while `Panic` is used for...
```

Author: This section evolved so much over time, I think it is due for a full rewrite at some point...

#. 0x41: If you allocate too much memory or create an array that is too large.
#. 0x51: If you call a zero-initialized variable of internal function type.

The ``require`` function either creates an error without any data,
an error of type ``Error(string)`` or a custom error type
and it
should be used to ensure valid conditions
that cannot be detected until execution time.
This includes conditions on inputs
or return values from calls to external contracts.

> A ``Error(string)`` exception (or an exception without data) is generated

Reviewer 3: ```suggestion
An ``Error(string)`` exception (or an exception without data) is generated
```

Reviewer 2: Yeah, I also think this should be `an`. I guess it could be `a` if someone reads it aloud in a way other than "an error string exception or ..."?

by the compiler
in the following situations:

> #. Calling ``require(x)`` where ``x`` evaluates to ``false``.

Reviewer 2: What about calling `revert()`? It is mentioned later in the text but I think it should be also on this list because it claims to have all the possible causes for `Error()`.

#. 0x41: If you allocate too much memory or create an array that is too large.
#. 0x51: If you call a zero-initialized variable of internal function type.

> The ``require`` function either creates an error without any data,

Reviewer: maybe: `without any data OR an error`

an error of type ``Error(string)`` and it
should be used to ensure valid conditions
that cannot be detected until execution time.
This includes conditions on inputs
or return values from calls to external contracts.

.. note::

    It is currently not possible to use custom errors in combination
    with ``require``. Please use ``if (!condition) revert CustomError();`` instead.

> An ``Error(string)`` exception (or an exception without data) is generated

Reviewer: I didn't checked yet the whole documentation, maybe its already described somewhere, however wouldn't it make sense to have a similar section for `Panic(uint256)`?

Reviewer 2: There already is one above this text. Starts with `A Panic exception is generated in the following situations.`

#. If you create a contract using the ``new`` keyword but the contract
   creation :ref:`does not finish properly<creating-contracts>`.

> You can optionally provide a message string or a custom error for ``require``, but not for ``assert``.

Reviewer 4: This is not currently true, is it?
```suggestion
You can optionally provide a message string for ``require``, but not for ``assert``.
```

.. note::

>     If you do not provide a string argument or custom error to ``require``, it will revert

Reviewer 4: Also for this.

    which consumed all gas available to the call.
    Exceptions that use ``require`` used to consume all gas until before the Metropolis release.

.. _revert-statement:

``revert``
----------

A direct revert can be triggered using the ``revert`` statement and the ``revert`` function.

The ``revert`` statement takes a custom error as direct argument without parentheses:

    revert CustomError(arg1, arg2);

> For backards-compatibility reasons, there is also the revert function, which uses parentheses

Reviewer 2: ```suggestion
For backards-compatibility reasons, there is also the ``revert()`` function, which uses parentheses
```

and accepts a string:

    revert();
    revert("description");

The error data will be passed back to the caller and can be caught there.

> Using ``revert()`` causes a revert without any error data while ``revert("description")``

Reviewer 2: I just realized that we do not allow `revert;` and for that use case the `revert()` function is still necessary. Maybe we should add it?

Author: Do we really want to allow reverts without error data?

Reviewer 2: I mean, we already do allow them, don't we? :)

So you mean it's because we want to discourage such reverts? Well, I'd say it depends whether people do use them. If they do, I think the new syntax should have an equivalent. But I don't remember seeing many bare reverts like this in Solidity code so maybe it's ok.

Author: I would wait and see if we have any requests. If there are none, we can remove the revert function with 0.9.0 altogether, otherwise we add an empty revert statement.

will create an ``Error(string)`` error.

> Using a custom error instance will usually be much cheaper than a string description,

Reviewer 4: ```suggestion
Using a custom error instance will usually be extremely cheaper than a string description,
```
Could use more emphasis, looking at the size of revert strings in contracts.

Reviewer: I would prefer `...will usually be cheaper than...`.  It is more neutral.

because you can use the name of the error to describe it, which is encoded in only

> four bytes. A longer description can be supplied via NatSpec which does not incur

Reviewer: instead of `incur`, i would prefer `increase` - its easier to understand.

Reviewer 2: I'd vote for `incur`. Seems natural to me in this context - it's often used in combination with `cost`.

The ``revert`` function is another way to trigger exceptions from within other code blocks to flag an error and
revert the current call. The function takes an optional string
message containing details about the error that is passed back to the caller

> and it will create an ``Error(string)`` exception, or you provide a custom error instance.

Reviewer 3: ```suggestion
and it will create either an ``Error(string)`` exception or an exception from the user-defined error type.
```

The ``revert`` function is another way to trigger exceptions from within other code blocks to flag an error and
revert the current call. The function takes an optional string
message containing details about the error that is passed back to the caller
and it will create an ``Error(string)`` exception, or you provide a custom error instance.

Using a custom error instance will usually be much cheaper than a string description.

The following example shows how to use an error string and a custom error instance
together with ``revert`` and the equivalent ``require``:

::

    // SPDX-License-Identifier: GPL-3.0

>     pragma solidity >=0.5.0 <0.9.0;

Reviewer 2: ```suggestion
    pragma solidity ^0.8.2;
```

            );
            // Perform the purchase.
        }
        function withdraw() public {
            if (msg.sender != owner) revert(Unauthorized());
            // Alternative way to do it:
            require(msg.sender == owner, Unauthorized());

            msg.sender.send(address(this).balance);
        }
    }

The two ways ``if (!condition) revert(...);`` and ``require(condition, ...);`` are

> equivalent as long as the argumenst to ``revert`` and ``require`` do not have side-effects,

Reviewer 2: ```suggestion
equivalent as long as the arguments to ``revert`` and ``require`` do not have side-effects,
```

            );
            // Perform the purchase.
        }
        function withdraw() public {
            if (msg.sender != owner)
                revert Unauthorized();

            msg.sender.send(address(this).balance);
        }
    }

The two ways ``if (!condition) revert(...);`` and ``require(condition, ...);`` are
equivalent as long as the arguments to ``revert`` and ``require`` do not have side-effects,

> for example if they are just strings or simple error instances.

Reviewer 4: About the error instance part: is that implemented now?

Author: Removed.
